lib/metric.js

```diff
@@ -1,51 +1,95 @@
 'use strict';
 
 const util = require('util');
-
-const metric = Symbol('metric');
-const source = Symbol('source');
-const notEmpty = Symbol('empty');
+const is = require('./is');
+
+const _description = Symbol('metric:key:description');
+const _timestamp = Symbol('metric:key:timestamp');
+const _source = Symbol('metric:key:source');
+const _labels = Symbol('metric:key:labels');
+const _value = Symbol('metric:key:value');
+const _time = Symbol('metric:key:time');
+const _meta = Symbol('metric:key:meta');
+const _type = Symbol('metric:key:type');
+const _name = Symbol('metric:key:name');
 
 const Metric = class Metric {
-    constructor(metricObj) {
-        this[metric] = metricObj;
-        this[source] = '';
+    constructor({
+        description,
+        timestamp = null,
+        source = null,
+        labels = [],
+        value = null,
+        time = null,
+        meta = null,
+        type = 0,
+        name,
+    } = {}) {
+        // Required arguments
+        if (!name) throw new Error('The argument "name" must be provided');
+        if (!description) throw new Error('The argument "description" must be provided');
+
+        // Validation
+        if (!is.validDescription(description)) throw new Error('Provided value to argument "description" is not legal');
+        if (!is.validTimestamp(timestamp)) throw new Error('Provided value to argument "timestamp" is not legal');
+        if (!is.validSource(source)) throw new Error('Provided value to argument "source" is not legal');
+        if (!is.validLabels(labels)) throw new Error('Provided value to argument "label" is not legal');
+        if (!is.validValue(value)) throw new Error('Provided value to argument "value" is not legal');
+        if (!is.validType(type)) throw new Error('Provided value to argument "type" is not legal');
+        if (!is.validName(name)) throw new Error('Provided value to argument "name" is not legal');
+
+        // Private properties
+        this[_description] = description;
+        this[_timestamp] = timestamp || Date.now() / 1000;
+        this[_source] = is.notEmpty(source) ? source : null;
+        this[_labels] = labels;
+        this[_value] = is.notEmpty(value) ? value : null;
+        this[_meta] = meta;
+        this[_type] = type;
+        this[_name] = name;
+
+        // Deprecated
+        this[_time] = is.notEmpty(time) ? time : null;
```

Comment:
should we use the deprecation pattern here?

Reply:
What thinking about that, but how will that play out? Either one will fill the logs with deprecation warnings because one need to do the check each time this object is instansiated. Or one will end up with doing the check once and risk not showing a deprecation warning for other places it might be in use.

Reply:
I think we're fine not deprecating it. Its not used that much yet.